tensorflow_model_optimization/python/core/clustering/keras/cluster.py

```diff
@@ -0,0 +1,213 @@
+from tensorflow.python.keras.engine.base_layer import Layer
+from tensorflow.python.keras.engine.input_layer import InputLayer
+from tensorflow.python.keras.utils.generic_utils import custom_object_scope
+
+from tensorflow_model_optimization.python.core.clustering.keras import clustering_centroids
+from tensorflow_model_optimization.python.core.clustering.keras.cluster_wrapper import ClusterWeights
+
+
+def clustering_scope():
+  """Provides a scope in which Clustered layers and models can be deserialized.
+
+  If a keras model or layer has been clustered, it needs to be within this scope
+  to be successfully deserialized.
+
+  Returns:
+      Object of type `CustomObjectScope` with clustering objects included.
+
+  Example:
+
+  ```python
+  clustered_model = cluster_weights(model, **self.params)
+  keras.models.save_model(clustered_model, keras_file)
+
+  with clustering_scope():
+    loaded_model = keras.models.load_model(keras_file)
+  ```
+  """
+  return custom_object_scope(
+    {
+      'ClusterWeights': ClusterWeights,
+    }
+  )
+
+
+def cluster_weights(to_cluster, number_of_clusters, cluster_centroids_init, **kwargs):
+  """Modify a keras layer or model to be clustered during training.
+
+  This function wraps a keras model or layer with clustering functionality
+  which clusters the layer's weights during training. For examples, using
+  this with number of clusters equals 8 will ensure that each weight tensor has
+  no more than 8 unique values.
+
+  Before passing to the clustering API, a model should already be trained and show some acceptable performance
+  on the testing/validation sets.
+
+  The function accepts either a single keras layer
+  (subclass of `keras.layers.Layer`), list of keras layers or a keras model
+  (instance of `keras.models.Model`) and handles them appropriately.
+
+  If it encounters a layer it does not know how to handle, it will throw an
+  error. While clustering an entire model, even a single unknown layer would lead
+  to an error.
+
+  Cluster a model:
+
+  ```python
+  clustering_params = {
+    'number_of_clusters': 8,
+    'cluster_centroids_init': 'density-based'
+  }
+
+  clustered_model = cluster_weights(original_model, **clustering_params)
+  ```
+
+  Cluster a layer:
+
+  ```python
+  clustering_params = {
+    'number_of_clusters': 8,
+    'cluster_centroids_init': 'density-based'
+  }
+
+  model = keras.Sequential([
+      layers.Dense(10, activation='relu', input_shape=(100,)),
+      cluster_weights(layers.Dense(2, activation='tanh'), **clustering_params)
+  ])
+  ```
+
+  Arguments:
+      to_cluster: A single keras layer, list of keras layers, or a
+        `tf.keras.Model` instance.
+      number_of_clusters: the number of cluster centroids to form when clustering a layer/model.
+        For example, if number_of_clusters=8 then only 8 unique values will be used in each weight array.
+      cluster_centroids_init: how to initialize the cluster centroids.
```

Comment:
Has there been evidence on which of the clustering init is best or most often the best? If so, it'd be useful to eventually provide a recommendation somewhere (e.g. tips in "Overview" page or here ... depending on the length).

Reply:
Out of the three current options, linear centroid initialization seems to be a preferred option. It is also in line with what is reported in the original paper linked in the description to this PR.

I agree that we should document insights like that. A separate "white-paper"-like document, or an "Overview" page would probably be a better option than the docstring.

We may also consider using appropriate default parameter values for the cases like this.